Map network type operations

austindrenski opened the pull request:
# Map network type operations

See #299 for more information.


Operator | Description | Example
-- | -- | --
`<` | is less than | `inet '192.168.1.5' < inet '192.168.1.6'`
`<=` | is less than or equal | `inet '192.168.1.5' <= inet '192.168.1.5'`
`=` | equals | `inet '192.168.1.5' = inet '192.168.1.5'`
`>=` | is greater or equal | `inet '192.168.1.5' >= inet '192.168.1.5'`
`>` | is greater than | `inet '192.168.1.5' > inet '192.168.1.4'`
`<>` | is not equal | `inet '192.168.1.5' <> inet '192.168.1.4'`
`<<` | is contained by | `inet '192.168.1.5' << inet '192.168.1/24'`
`<<=` | is contained by or equals | `inet '192.168.1/24' <<= inet '192.168.1/24'`
`>>` | contains | `inet '192.168.1/24' >> inet '192.168.1.5'`
`>>=` | contains or equals | `inet '192.168.1/24' >>= inet '192.168.1/24'`
`&&` | contains or is contained by | `inet '192.168.1/24' && inet '192.168.1.80/28'`
`~` | bitwise NOT | `~ inet '192.168.1.6'`
`&` | bitwise AND | `inet '192.168.1.6' & inet '0.0.0.255'`
`\|` | bitwise OR | `inet '192.168.1.6' \| inet '0.0.0.255'`
`+` | addition | `inet '192.168.1.6' + 25`
`-` | subtraction | `inet '192.168.1.43' - 36`
`-` | subtraction | `inet '192.168.1.43' - inet '192.168.1.19'`

Function | Description 
-- | -- 
`abbrev(inet)` | abbreviated display format as text 
`abbrev(cidr)` | abbreviated display format as text 
`broadcast(inet)` | broadcast address for network 
`family(inet)` | extract family of address; 4 for IPv4, 6 for IPv6 
`host(inet)` | extract IP address as text 
`hostmask(inet)` | construct host mask for network 
`masklen(inet)` | extract netmask length 
`netmask(inet)` | construct netmask for network 
`network(inet)` | extract network part of address 
`set_masklen(inet, int)` | set netmask length for inet value 
`set_masklen(cidr, int)` | set netmask length for cidr value 
`text(inet)` | extract IP address and netmask length as text 
`inet_same_family(inet, inet)` | are the addresses from the same family? 
`inet_merge(inet, inet)` | the smallest network which includes both of the given networks 
`trunc(macaddr)` | set last 3 bytes to zero 
`trunc(macaddr8)` | set last 5 bytes to zero 
`macaddr8_set7bit(macaddr8)` | set 7th bit to one, also known as modified EUI-64, for inclusion in an IPv6 address

roji commented:
@austindrenski, I know this PR is still in progress, but here's one thought... Unlike with the range operators, I think the inet operations shouldn't exist as a set of extension methods over `IPAddress`, but rather as extension methods on `DbFunctions`, requiring users to invoke them explicitly via `EF.Functions.Contains()`.

The reason for this different is that `IPAddress` is a general .NET type that isn't used exclusively for database access, unlike ranges or the full text search types. It doesn't sound like a good idea for users to suddenly have `IPAddress.Contains()` appear just because they reference the Npgsql provider (not to mention that the method throws anywhere except for EF Core queries).

roji commented:
> My thought process was that if local implementations can be provided, and if those implementations are consistent with PostgreSQL, then extension methods would make sense.

I'm still very wary of adding extensions methods on standard, non-database-related types in an EF Core provider... A good example is the [`Like`](https://github.com/aspnet/EntityFrameworkCore/blob/76d0d43c3757ef1e0d69ece840d76bb3bca201be/src/EFCore/DbFunctionsExtensions.cs#L36) function provided by EF Core, which has a C# implementation but is still accessible as `EF.Functions.Like()` rather than as an extension method over string. The fear is that people start using the function without even being aware of where it's coming from (that's the danger of extensions methods), and at the end of the day we're a database provider, and not a general functionality provider...

(if you're interested we had a conversation about this with the EF Core team, I can dig it up).

> On the technical side, I'm hoping that PostgreSQL network address operations will be easier to replicate than range operators. I could be wrong here. But as in #323, inconsistent local implementations would be worse than always-throw implementations.

I agree... I tend towards throwing in general, since IMHO there isn't a lot of value in the client implementation, and we have to be sure to provide a 100% accurate (and reasonably efficient) implementation - I'm not sure this is really worth our time. At the end of the day it's really up to you, but I'd recommend against it.

> I noticed that NpgsqlInet doesn't have an entry in NpgsqlTypeMappingSource.ClrTypeMappings, but there is a cidr mapping in NpgsqlTypeMappingSource.StoreTypeMappings.
>
> For testing, I've added an entry for on this branch and it seems to work as expected.
> 
> Was this an oversight? Or is there a reason not to include the mapping?

It's a bit of an oversight. Right now the so-called "provider-specific types" (NpgsqlInet, NpgsqlDateTime and the others) really supported in the EF Core provider - they're quite rarely-used. In most cases they only exist to provide access to the full range of precision/scale of the PostgreSQL type where the CLR type is more limited (e.g. PostgreSQL `timestamp` has 1 microsecond precision, but CLR `DateTime` has 10 microsecond precision). NpgsqlInet is a bit special because it bundles both an IP and a subnet, which is something that doesn't exist at all in CLR (so not just a precision difference).

Long story short, yes, I think it makes sense to have the mappings as follows:

* `IPAddress` - `inet` (default CLR mapping)
* `NpgsqlInet` -> `cidr` (default CLR mapping)
* `cidr` -> `NpgsqlInet` (default store mapping
* `inet` -> `IPAddress` (default store mapping)
* `inet` -> `NpgsqlInet` (non-default store mapping, for people who want to use `inet` to represent a network - this is supported by PostgreSQL).

roji commented:
>This may be a silly question, but instead of default mapping cidr and NgpsqlInet, what do you think about simply setting the default mapping to a ValueTuple<IPAddress, int>?
>
> It feels complicated to have one network type (inet) map to a built-in type, while a closely related network type (cidr) maps to a "provider-specific type".

I really like this idea. NpgsqlInet is almost the equivalent of an IPAdress/int tuple (it's a struct), it was created way before tuples existed.

The only catch is that this would need to be done at the ADO.NET level first and foremost. That's no big deal, I can do that soon, and at that point you can submit a PR for the corresponding changes in EF Core. Sounds good?

> On a related note, does mapping inet to IPAddress pose a risk for data loss/truncation?

You mean because PostgreSQL `inet` can contain a subnet and `IPAddress` can't? That's true, but since we really had have two PG types, one really meant for doing networks (`cidr`) and mainstream use of `inet` is to hold a single IP, it seems like a shame to encumber the common simple IP scenario. Plus you always have the option of explicitly mapping to the tuple. So I think it should be ok...

austindrenski commented:
## Type mapping

> The only catch is that this would need to be done at the ADO.NET level first and foremost. That's no big deal, I can do that soon, and at that point you can submit a PR for the corresponding changes in EF Core. Sounds good?

That works for me!

> since we really had have two PG types, one really meant for doing networks (cidr) and mainstream use of inet is to hold a single IP, it seems like a shame to encumber the common simple IP scenario. Plus you always have the option of explicitly mapping to the tuple. So I think it should be ok...

I agree that it would impose a burden on the common IP use... But it does feel like the makings of a [footgun](https://en.wiktionary.org/wiki/footgun) situation... Would it introduce too much overhead to log some type of warning if this occurs during materialization? 

For example, ___if___ the materialization is mapping `inet` -> `IPAddress` ___and___ it has subnet info, ___then___ log a warning of possible data loss.

If that's too much overhead (in general or relative to the risk), perhaps we can highlight the risk in the docs?

## Question on parameter reuse and duplication

I'm seeing parameter duplication when the same `NpgsqlInet` is used multiple times in a query. There's also some convoluted equality handling...but that looks unrelated to the parameter duplication.

Is this expected behavior? Related to value types being copied on closure? Something else entirely?

__Example:__

```c#
[Fact]
public void Demonstrate_ValueTypeParametersAreDuplicated()
{
    using (NetContext context = Fixture.CreateContext())
    {
        NpgsqlInet npgsqlInet = new IPAddress(0);

        bool[] _ =
            context.NetTestEntities
                   .Where(x => x.CidrMappedToNpgsqlInet.ContainsOrEquals(npgsqlInet))
                   .Select(x => x.CidrMappedToNpgsqlInet.Equals(npgsqlInet))
                   .ToArray();

        AssertContainsSql("SELECT x.\"CidrMappedToNpgsqlInet\" == @__npgsqlInet_0");
        AssertContainsSql("WHERE x.\"CidrMappedToNpgsqlInet\" >>= @__npgsqlInet_0");
    }
}
```

__Expectation:__

```sql
SELECT x."CidrMappedToNpgsqlInet" = @__npgsqlInet_0
FROM "NetTestEntities" AS x
WHERE x."CidrMappedToNpgsqlInet" >>= @__npgsqlInet_0 = TRUE
```

__Reality:__

```sql
SELECT CASE
    WHEN x."CidrMappedToNpgsqlInet" = @__npgsqlInet_1
    THEN TRUE::bool ELSE FALSE::bool
END
FROM "NetTestEntities" AS x
WHERE x."CidrMappedToNpgsqlInet" >>= @__npgsqlInet_0 = TRUE
```

roji commented:
See #1937 for a PR for your changes at the ADO level, let me know what you think.

> On a related note, does mapping inet to IPAddress pose a risk for data loss/truncation?

Regarding the data loss/truncation problem, I don't think there's actually a problem. [The PostgreSQL docs](https://www.postgresql.org/docs/current/static/datatype-net-types.html) say that "The inet type holds an IPv4 or IPv6 host address, and optionally its subnet". This is also shown by the fact that PostgreSQL doesn't force you to specify a subnet in the `inet` literal representation - you can put a simple IP and that's OK.

From the .NET perspective, it's a mater of what the user tells us they want. They always the option of saying they want the subnet - this is done by requesting an `NpgsqlInet`, or the tuple in the new proposed API. If they choose to read only an address, they are discarding the subnet part of it. This also happens to be the default behavior for the `inet` type, but I don't think that changes anything.

Note that this is very different from precision truncation, where, say, a database double loses precision when read... This is a column containing two distinct pieces of information, one of which is totally optional. So I think it's OK (maybe @YohDeadfall has an opinion).

I don't have time at the moment to look at the parameter reuse/duplication issue above, I'll try to do so soon.

austindrenski commented:
@roji An exception is thrown when using the new mapping from `cidr` to `(IPAddress IPAddress, int Subnet)`.

Should I bump Npgsql from `4.0.0-ci.0` to `4.0.0.0-ci.931`?

```
System.InvalidCastException : Can't write CLR type System.ValueTuple`2[System.Net.IPAddress,System.Int32] with handler type CidrHandler
   at lambda_method(Closure , NpgsqlTypeHandler , Object , NpgsqlLengthCache& , NpgsqlParameter )
   at Npgsql.TypeHandling.NpgsqlSimpleTypeHandler`1.ValidateObjectAndGetLength(Object value, NpgsqlLengthCache& lengthCache, NpgsqlParameter parameter) in C:\projects\npgsql\src\Npgsql\TypeHandling\NpgsqlSimpleTypeHandler.cs:line 236
   at Npgsql.NpgsqlParameter.ValidateAndGetLength() in C:\projects\npgsql\src\Npgsql\NpgsqlParameter.cs:line 557
   at Npgsql.NpgsqlCommand.ValidateParameters() in C:\projects\npgsql\src\Npgsql\NpgsqlCommand.cs:line 795
   at Npgsql.NpgsqlCommand.ExecuteDbDataReader(CommandBehavior behavior, Boolean async, CancellationToken cancellationToken) in C:\projects\npgsql\src\Npgsql\NpgsqlCommand.cs:line 1143
   at Npgsql.NpgsqlCommand.ExecuteDbDataReader(CommandBehavior behavior) in C:\projects\npgsql\src\Npgsql\NpgsqlCommand.cs:line 1132
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, DbCommandMethod executeMethod, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteReader(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.BufferlessMoveNext(DbContext _, Boolean buffer)
   at Npgsql.EntityFrameworkCore.PostgreSQL.Storage.Internal.NpgsqlExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded) in C:\Users\austin.drenski\Rider\Npgsql.EntityFrameworkCore.PostgreSQL\src\EFCore.PG\Storage\Internal\NpgsqlExecutionStrategy.cs:line 51
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider._TrackEntities[TOut,TIn](IEnumerable`1 results, QueryContext queryContext, IList`1 entityTrackingInfos, IList`1 entityAccessors)+MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
   at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at Npgsql.EntityFrameworkCore.PostgreSQL.Query.NetworkAddressQueryNpgsqlTest.NpgsqlInetContainsNpgsqlInet() in C:\Users\austin.drenski\Rider\Npgsql.EntityFrameworkCore.PostgreSQL\test\EFCore.PG.FunctionalTests\Query\NetworkAddressQueryNpgsqlTest.cs:line 87
```

austindrenski commented:
I rebased the PR to the current dev and updated some NuGet settings to target the latest Npgsql build.

However, an exception is thrown if I attempt to construct test entities to add and save into the test context... If I comment out the local construct/add/save code, then the unit tests go back to passing. So the CLR to Postgres translation seems to be working fine.

It looks like the stack trace is complaining that [tuples lack an `==` operator](https://github.com/dotnet/csharplang/issues/190). Does Npgsql (or EF Core) require an `==` operator for something?

Do you have any insight into what might be causing this, @roji?

```
Npgsql.EntityFrameworkCore.PostgreSQL.Query.NetworkAddressQueryNpgsqlTest.IPAddressContainOrEqualIPAddress

System.AggregateException : One or more errors occurred. (The binary operator Equal is not defined for the types 'System.ValueTuple`2[System.Net.IPAddress,System.Int32]' and 'System.ValueTuple`2[System.Net.IPAddress,System.Int32]'.) (The following constructor parameters did not have matching fixture data: NetworkAddressQueryNpgsqlFixture fixture)
---- System.InvalidOperationException : The binary operator Equal is not defined for the types 'System.ValueTuple`2[System.Net.IPAddress,System.Int32]' and 'System.ValueTuple`2[System.Net.IPAddress,System.Int32]'.
---- The following constructor parameters did not have matching fixture data: NetworkAddressQueryNpgsqlFixture fixture

----- Inner Stack Trace #1 (System.InvalidOperationException) -----
   at Microsoft.EntityFrameworkCore.Metadata.Internal.ClrAccessorFactory`1.Create(PropertyInfo propertyInfo, IPropertyBase propertyBase)
   at Microsoft.EntityFrameworkCore.Internal.NonCapturingLazyInitializer.EnsureInitialized[TParam,TValue](TValue& target, TParam param, Func`2 valueFactory)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.ReadPropertyValue(IPropertyBase propertyBase)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.ChangeDetector.DetectChanges(InternalEntityEntry entry)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.ChangeDetector.DetectChanges(IStateManager stateManager)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChanges(Boolean acceptAllChangesOnSuccess)
   at Npgsql.EntityFrameworkCore.PostgreSQL.Query.NetworkAddressQueryNpgsqlTest.NetworkAddressQueryNpgsqlFixture..ctor() in C:\Users\austin.drenski\Rider\Npgsql.EntityFrameworkCore.PostgreSQL\test\EFCore.PG.FunctionalTests\Query\NetworkAddressQueryNpgsqlTest.cs:line 270
----- Inner Stack Trace #2 (Xunit.Sdk.TestClassException) -----
```

austindrenski commented:
## Updates
eb491208197ec9c9c95c0d30d32dc1b25196c85f completes the [operators](https://www.postgresql.org/docs/current/static/functions-net.html#CIDR-INET-OPERATORS-TABLE)  for `inet` and `cidr`.

I was adding exception messages to each function...but that started to look like a maintenance nightmare. So instead, I defined `ClientEvalutaionNotSupportedException` which extends `NotSupportedException` with a constructor that uses `CallerMemberNameAttribute` to forward the method's name to a templated message.

I had considered adding overloads for passing one `inet` and one `cidr`, but I think we can skip that since `cidr` is now explicitly a tuple of `IPAddress` and `int`. That should make it obvious (and trivial) for callers to mix and match `inet` and `cidr` for functions.

## Next steps

- I need to finish the series of functional tests for the operators. 

- I'm planning to start on the [functions](https://www.postgresql.org/docs/current/static/functions-net.html#CIDR-INET-FUNCTIONS-TABLE) for `inet` and `cidr`. 
   - Since we're nearing `4.0.0`, do you want the function work split into another PR?

- I have also been __unable__ to resolve [the exception](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/pull/407#issuecomment-391838307) that occurs when adding `cidr` values to the test database.

roji commented on src/EFCore.PG/Extensions/NpgsqlNetworkExtensions.cs:
```diff
@@ -0,0 +1,1006 @@
+        /// <exception cref="ClientEvaluationNotSupportedException">
+        /// This method is only intended for use via SQL translation as part of an EF Core LINQ query.
+        /// </exception>
+        public static bool LessThan([CanBeNull] this DbFunctions _, IPAddress inet, IPAddress other) => throw new ClientEvaluationNotSupportedException();
```
nit: please put the `=> throw ...` on its own line to reduce wide lines

roji commented on src/EFCore.PG/Extensions/NpgsqlNetworkExtensions.cs:
```diff
@@ -0,0 +1,1006 @@
+
+        #endregion
+
+        #region ContainmentOperators
```
Aren't we missing the check of containment of `inet` in a `cidr`, which seems to be the most relevant one? Or does PostgreSQL not support this somehow? Same for some of the other operations here.

austindrenski replied:
The implicit casting between `inet` and `cidr` makes it hard to decide when to stop including overloads, but I think you're right on the containment checks being important. 

I've added a few more in the next commit. Happy to add more, just let me know if there are others you want to see included.

roji commented on src/EFCore.PG/Extensions/NpgsqlNetworkExtensions.cs:
```diff
@@ -0,0 +1,1006 @@
+        /// <exception cref="ClientEvaluationNotSupportedException">
+        /// This method is only intended for use via SQL translation as part of an EF Core LINQ query.
+        /// </exception>
+        public static IPAddress Not([CanBeNull] this DbFunctions _, IPAddress inet) => throw new ClientEvaluationNotSupportedException();
```
I propose to rename to `BitwiseNot` to make things more clear (same for the other bit operations below), a "not" operation on an IP address is far from being an everyday thing. This is also how it is documented in the PostgreSQL docs.

roji commented on src/EFCore.PG/Extensions/NpgsqlNetworkExtensions.cs:
```diff
@@ -0,0 +1,1006 @@
+
+        #endregion
+
+        #region Functions
```
I realize that the `cidr` tuple has `Subnet`, but I think the methods below should be named as close as possible to ther PostgreSQL equivalents (aside from obvious casing differences with C#), so I'd retain `MaskLength()` rather than `SubnetLength()` etc.

roji commented on src/EFCore.PG/Extensions/NpgsqlRangeExtensions.cs:
```diff
@@ -32,6 +33,7 @@ namespace Microsoft.EntityFrameworkCore
     /// <summary>
     /// Provides extension methods for <see cref="NpgsqlRange{T}"/> supporting PostgreSQL translation.
     /// </summary>
+    [PublicAPI]
```
We don't really use this Jetbrains annotation, it's truly you'll find it here and there but I'd rather avoid having it on every actual public method etc.

roji commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlNetworkTranslator.cs:
```diff
@@ -0,0 +1,148 @@
+                return new CustomBinaryExpression(expression.Arguments[1], expression.Arguments[2], "-", expression.Arguments[1].Type);
+
+            case nameof(NpgsqlNetworkExtensions.Abbreviate):
+                return new PgFunctionExpression("abbrev", typeof(string), new[] { expression.Arguments[1] });
```
Please use `SqlFunctionExpression` instead of `PgFunctionExpression`, the latter is there only for named arguments (which is a PostgreSQL-specific syntax) which isn't used here.

roji commented on src/EFCore.PG/Storage/Internal/Mapping/NpgsqlNetworkTypeMappings.cs:
```diff
@@ -1,8 +1,30 @@
-﻿using System;
-using System.Collections.Generic;
+﻿#region License
```
Changes in this file were already submitted via #429

roji commented on src/EFCore.PG/Utilities/ClientEvaluationNotSupportedException.cs:
```diff
@@ -0,0 +1,48 @@
+    /// <summary>
+    /// The exception that is thrown when a method intended for SQL translation is evaluated by the client.
+    /// </summary>
+    public class ClientEvaluationNotSupportedException : NotSupportedException
```
This is a good idea, but it belongs in EFCore rather than in the PostgreSQL provider, can you please submit a PR to them for this?

In the meantime we can still use this.

austindrenski commented:
@roji 6f881fd57eb861040fe7be4da0f3a76a9587f01c rebases and attempts to address your review.

I think the open question is whether there are a sufficient mix of extension overloads. What do you think?

As an aside, this makes a nice example of how something like https://github.com/dotnet/csharplang/issues/515 could really simplify things...

```c# 
public static class Extensions
{
    public static explicit operator (IPAddress Address, Subnet int)(IPAddress address)
    {
        return (address, 0);
    }

    public static explicit operator IPAddress((IPAddress Address, Subnet int) network)
    {
        return network.Address;
    }
}
```

austindrenski commented:
Rebased to dev for the latest tests.

austindrenski commented:
@roji With the decision in #428, I think that this PR is complete on my end. 

Are there any further issues/features that you would like me to address before the merge?

austindrenski commented:
@roji This is updated and ready for another look.

austindrenski commented:
@roji I've given this another pass after #529. At this point, I think it's about as lean as it can be.

Any chance you have time to give it a last look? I would like to get it merged today, before I start breaking #431 into smaller PRs.

roji commented on src/EFCore.PG/Extensions/NpgsqlNetworkExtensions.cs:
```diff
@@ -0,0 +1,1180 @@
+        /// This method is only intended for use via SQL translation as part of an EF Core LINQ query.
+        /// </exception>
+        public static bool LessThan([CanBeNull] this DbFunctions _, IPAddress inet, IPAddress other)
+            => throw ClientEvaluationNotSupportedException();
```
As discussed elsewhere, let's throw the standard `NotSupportedException` rather than this.

austindrenski replied:
This is actually just a [helper method](https://github.com/austindrenski/Npgsql.EntityFrameworkCore.PostgreSQL/blob/3ca6977533dfd7b8475e678ed1d681d8f3044992/src/EFCore.PG/Extensions/NpgsqlNetworkExtensions.cs#L1175-L1176) now:

```c#
static NotSupportedException ClientEvaluationNotSupportedException(
    [CallerMemberName] string method = default)
    => new NotSupportedException($"{method} is only intended for use via SQL translation as part of an EF Core LINQ query.");
```

roji replied:
Ohh ok...

roji commented on src/EFCore.PG/Extensions/NpgsqlNetworkExtensions.cs:
```diff
@@ -0,0 +1,1180 @@
+        /// <exception cref="NotSupportedException">
+        /// This method is only intended for use via SQL translation as part of an EF Core LINQ query.
+        /// </exception>
+        public static bool Equal([CanBeNull] this DbFunctions _, IPAddress inet, IPAddress other)
```
Thinking about this again, this isn't necessary... EF Core already translates `x.Equals(y)` to `x = y` in SQL.

So it should be fine to remove all `Equal` and `NotEqual` overloads below.

roji commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlNetworkTranslator.cs:
```diff
@@ -0,0 +1,149 @@
+            case nameof(NpgsqlNetworkExtensions.LessThanOrEqual):
+                return new CustomBinaryExpression(expression.Arguments[1], expression.Arguments[2], "<=", typeof(bool));
+
+            case nameof(NpgsqlNetworkExtensions.Equal):
```
As above, remove.

roji commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlNetworkTranslator.cs:
```diff
@@ -0,0 +1,149 @@
+            case nameof(NpgsqlNetworkExtensions.GreaterThan):
+                return new CustomBinaryExpression(expression.Arguments[1], expression.Arguments[2], ">", typeof(bool));
+
+            case nameof(NpgsqlNetworkExtensions.NotEqual):
```
As above, remove.

roji commented on test/EFCore.PG.FunctionalTests/Query/NetworkQueryNpgsqlTest.cs:
```diff
@@ -0,0 +1,1773 @@
+        /// <summary>
+        /// Asserts that the SQL fragment appears in the logs.
+        /// </summary>
+        /// <param name="sql">The SQL statement or fragment
```
nit: expression-bodied method

roji commented on test/EFCore.PG.FunctionalTests/Query/NetworkQueryNpgsqlTest.cs:
```diff
@@ -0,0 +1,1773 @@
+            /// A <see cref="NetContext"/> for testing.
+            /// </returns>
+            public NetContext CreateContext()
+            {
```
nit: expression-bodied method here and elsewhere

austindrenski commented:
@roji Just pushed d4d851fd2e1293217340d78868cb7956fd048429 to address your review.

roji commented:
Thanks, LGTM.

Please squash everything, leaving a single concise commit...
